lib/dalli/protocol/base.rb

```diff
@@ -201,13 +211,14 @@ def connect
         raise
       end
 
-      def pipelined_get(keys)
-        req = +''
-        keys.each do |key|
-          req << quiet_get_request(key)
-        end
-        # Could send noop here instead of in pipeline_response_setup
-        write(req)
+      def pipelined_get(bytes)
+        write_nonblock(bytes)
+      rescue SystemCallError, Timeout::Error, EOFError => e
+        @connection_manager.error_on_request!(e)
+      end
+
+      def pipelined_get_capacity(keys)
+        (keys.size * request_header_size) + keys.reduce(0) { |acc, k| acc + k.size }
```

Comment:
```suggestion
        (keys.size * request_header_size) + keys.map(&:size).sum
```

In some quick benchmarking this looks 12-15% faster maybe, and I think more readable